### ISSUE TITLE
Connection field complexity calculation

### ISSUE THREAD

**Issue by crpahl** [rmosolgo/graphql-ruby#3755]
This is a follow-up to https://github.com/rmosolgo/graphql-ruby/issues/3591

After we upgraded to `1.13.0`, we noticed that the complexity computed by the `QueryComplexity` analyzer wasn't matching the actual complexity of the query when connection fields were queried (we track the actual complexity in a tracer which always gives us the actual complexity cost). It looks like it's because `nodes` and `edges` are getting included in the multiplication of `items_complexity` [here.](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/schema/field.rb#L447)

But `edges` and `nodes` should just count for 1 (or whatever their configured complexity is). A simple fix to get the correct query cost was to change it to this:

```
              metadata_complexity = 0
              nodes_edges_complexity = 0
              lookahead = GraphQL::Execution::Lookahead.new(query: query, field: self, ast_nodes: nodes, owner_type: owner)

              if (page_info_lookahead = lookahead.selection(:page_info)).selected?
                metadata_complexity += 1 # pageInfo
                metadata_complexity += page_info_lookahead.selections.size # subfields
              end

              if lookahead.selects?(:total) || lookahead.selects?(:total_count) || lookahead.selects?(:count)
                metadata_complexity += 1
              end

              nodes_edges_complexity += 1 if lookahead.selects?(:edges)
              nodes_edges_complexity += 1 if lookahead.selects?(:nodes)

              items_complexity = child_complexity - metadata_complexity - nodes_edges_complexity
              # Add 1 for _this_ field
              1 + (max_possible_page_size * items_complexity) + metadata_complexity + nodes_edges_complexity
```

After looking a little more closely at the complexity calculation code, I was wondering if we could do something like this instead...

Add the ability to store the computed child complexity on a field after it's been computed and add this to `GraphQL::Analysis::AST::QueryComplexity`:

```
            field_complexity(complexity_for_key, max_complexity: max_complexity, child_complexity: child_complexity)
            complexity_for_key.child_complexity = child_complexity
```

Add this method to `ScopedTypeComplexity`:

```
          def child_complexity=(child_complexity)
            @field_definition.child_complexity = child_complexity
          end
```

And then the complexity calculation in `calculate_complexity` can be updated to this:

```
              nodes_edges_complexity = 0

              lookahead = GraphQL::Execution::Lookahead.new(query: query, field: self, ast_nodes: nodes, owner_type: owner)
              nodes_edges_complexity += lookahead.selection(:edges).field.child_complexity if lookahead.selects?(:edges)
              nodes_edges_complexity += lookahead.selection(:nodes).field.child_complexity if lookahead.selects?(:nodes)

              # Add 1 for _this_ field
              1 + (max_possible_page_size * nodes_edges_complexity) + (child_complexity - nodes_edges_complexity)
```

Since the child complexity of all subfields has already been computed in the analyzer, we just need a way to access them here. I tried these changes out and it worked as expected, but this code is still new to me so I'd love to hear your thoughts!

**Comment by rmosolgo** [comment]
Hey, thanks for reporting this. 

> ```
> @field_definition.child_complexity = ... 
> ```

`@field_definition` here is a single `GraphQL::Schema::Field` instance shared by all threads in the process. In a multi-threaded server (like Puma, the Rails default), writes to that instance variable will clobber each other and threads might read different values than they previously wrote. It might work in development, where there's only a single thread, and with a process-based server like Unicorn, you might even get away with it in production. 

I'm game to investigate an improvement to that part of the code, but it'll need some more local place to store those calculations. 

In the meantime, I've worked up the other fix you suggested here: https://github.com/rmosolgo/graphql-ruby/pull/3758 Thanks for pointing it out!

**Comment by rmosolgo** [comment]
I _think_ this was fixed by #3758, but if I misunderstood, please let me know :+1:
